src/AppInstallerRepositoryCore/Microsoft/CheckpointIndex.cpp

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "CheckpointIndex.h"
+#include "SQLiteStorageBase.h"
+#include "Schema/Checkpoint_1_0/CheckpointIndexInterface.h"
+
+namespace AppInstaller::Repository::Microsoft
+{
+    namespace
+    {
+        constexpr std::string_view s_Checkpoints = "Checkpoints"sv;
+    }
+
+    CheckpointIndex CheckpointIndex::CreateNew(const std::string& filePath, Schema::Version version)
+    {
+        AICLI_LOG(Repo, Info, << "Creating new Checkpoint Index with version [" << version << "] at '" << filePath << "'");
+        CheckpointIndex result{ filePath, version };
+
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(result.m_dbconn, "CheckpointIndex_createnew");
+
+        // Use calculated version, as incoming version could be 'latest'
+        result.m_version.SetSchemaVersion(result.m_dbconn);
+
+        result.m_interface->CreateTables(result.m_dbconn);
+
+        result.SetLastWriteTime();
+
+        savepoint.Commit();
+
+        return result;
+    }
+
+    std::shared_ptr<CheckpointIndex> CheckpointIndex::OpenOrCreateDefault(GUID guid, OpenDisposition openDisposition)
+    {
+        const auto& indexPath = GetCheckpointIndexPath(guid);
+        AICLI_LOG(Repo, Info, << "Opening checkpoint index");
+
+        try
+        {
+            if (std::filesystem::exists(indexPath))
+            {
+                if (std::filesystem::is_regular_file(indexPath))
+                {
+                    try
+                    {
+                        AICLI_LOG(Repo, Info, << "Opening existing checkpoint index");
+                        return std::make_shared<CheckpointIndex>(CheckpointIndex::Open(indexPath.u8string(), openDisposition));
+                    }
+                    CATCH_LOG();
+                }
+
+                AICLI_LOG(Repo, Info, << "Attempting to delete bad checkpoint index file");
+                std::filesystem::remove_all(indexPath);
+            }
+
+            return std::make_shared<CheckpointIndex>(CheckpointIndex::CreateNew(indexPath.u8string()));
+        }
+        CATCH_LOG();
+
+        return {};
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    std::optional<std::filesystem::path> s_CheckpointIndexDirectoryOverride{};
+    void TestHook_SetCheckpointIndexDirectory_Override(std::optional<std::filesystem::path>&& checkpointIndexDirectory)
+    {
+        s_CheckpointIndexDirectoryOverride = std::move(checkpointIndexDirectory);
+    }
+#endif
+
+    std::filesystem::path CheckpointIndex::GetCheckpointIndexPath(GUID guid)
+    {
+        wchar_t checkpointGuid[256];
+        THROW_HR_IF(E_UNEXPECTED, !StringFromGUID2(guid, checkpointGuid, ARRAYSIZE(checkpointGuid)));
+
+        const auto DefaultIndexDirectoryPath = Runtime::GetPathTo(Runtime::PathName::LocalState) / s_Checkpoints;
```

Comment:
Add a new `Runtime` path, then use the existing runtime path test hooks to change it as needed.